### ISSUE TITLE
Needed function is lacking in reference.h

### ISSUE THREAD

**Issue by simpleyo** [godotengine/godot#9480]
**Operating system or device - Godot version:**
master branch.

In core/reference.h, in class Ref,  is needed the function:

	_FORCE_INLINE_ T* ptr() const {

		return reference;
	}

Without that function i can´t get a T* from a const Ref.

**Comment by bojidar-bg** [comment]
Well, you can get a const T* already, as per [line 131](https://github.com/godotengine/godot/blob/40bb90fabdbb98af9235b224d2172e4c1aeeacc9/core/reference.h#L131). If you really think you need a non-const pointer (why is in a const ref then, anyway?) you can cast the result of that function manually. See [this SO question for an example](https://stackoverflow.com/questions/8570139/is-a-cast-from-pointer-to-const-to-pointer-to-non-const-invalid-c).

Note that if you really do modify a const object, then, you are actively working around const-ness, which is probably there for a reason.

**Comment by simpleyo** [comment]
An example:

```

class A : public Node2D {

	OBJ_TYPE( A, Node2D );

    Ref<Texture>    texture;
    
public:

    // I can get a Ref<Texture> from a const A object
    Ref<Texture> get_texture() const { return texture; }  
    
    // I can't get a Texture* from a const A object    
    Texture* get_texture_ptr() const { return texture.____; }  
    
};

```

**why is in a const ref then, anyway?**

Is const because i want to use  get_texture_ptr() const because the texture object itself is not owned by A (A owns a reference to the texture object. You can see that i can modify the texture object, of a const A object, by means of get_texture() const that give me a Ref to a modificable object. Then, why is not a T* ptr() const method in class Ref?  I do not see any rational explanation.

**Comment by Zylann** [comment]
Can you use `*texture_ref` to get your `Texture*`?

**Comment by simpleyo** [comment]
No.

There is no  `T* operator->() const` method in class Ref. The same happens for `T* ptr() const` method.

**Comment by bojidar-bg** [comment]
Can't you just do `((A*)this).texture.ptr()` or `(Texture*)(texture.ptr())`?

**Comment by Zylann** [comment]
But... I do *ref all the time? (here as a reference) https://github.com/godotengine/godot/blob/master/editor/plugins/curve_editor_plugin.cpp#L376

**Comment by simpleyo** [comment]
Get a non-const pointer from a const Ref is not possible with the current implementation of Ref. For example, std::shared_ptr has a function` T* get() const` (note that is not `const T* get() const`).. With that it is possible to get a non-const pointer from a const std::shared_ptr. The question remains: Why `const T* ptr() const `and `const T* operator*() const` return a `const T*` and not a `T*`?. Can someone answer that question? I think there is no reason to return a const T* instead a T*.

**Comment by Calinou** [comment]
Is still relevant in Godot 3.2?

**Comment by Zylann** [comment]
You still can't convert a const Ref<T> into a T*.

**Comment by KoBeWi** [comment]
In either case it's more a missing feature, so it should go to proposals:
https://github.com/godotengine/godot-proposals

**Comment by bojidar-bg** [comment]
Actually, it seems to have been fixed in #64789: https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/object/ref_counted.h#L108

(Also, looking back at my comments, please accept my apologies for not understanding the problem initially. Now that I've programmed in Rust, I can understand what interior mutability is all about, but back then I had no idea about that whole world...)
